File: 0231_Power_of_Two/Solution2.py
class Solution:
    def isPowerOfTwo(self, n: int) -> bool:
        return n > 0 and ((n & (n-1)) == 0)

def test(test_name, n, expected):
    res = Solution().isPowerOfTwo(n)
    if res == expected:
        print(test_name + ' success.')
    else:
        print(test_name + ' failed.')


if __name__ == "__main__":
    n1, expected1 = 1, True
    test('test1', n1, expected1)

    n2, expected2 = 16, True
    test('test2', n2, expected2)

    n3, expected3 = 218, False
    test('test3', n3, expected3)

    n4, expected4 = 0, False
    test('test4', n4, expected4)

# Given an integer, write a function to determine if it is a power of two.

# Example 1:

# Input: 1
# Output: true 
# Explanation: 20 = 1

# Example 2:

# Input: 16
# Output: true
# Explanation: 24 = 16

# Example 3:

# Input: 218
# Output: false

File: 面试题14-I_剪绳子/Solution2.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

class Solution {
public:
    int cuttingRope(int n) {
        if (n < 4)      // n = 2, 3
            return n-1;
        int a = n / 3, b = n % 3;
        if (b == 0)
            return pow(3, a);
        if (b == 1)
            return pow(3, a-1) * 4;
        if (b == 2)
            return pow(3, a) * 2;
        return 0;                   // impossible, for compile
    }
};

void test(string test_name, int n, int expected)
{
    int res = Solution().cuttingRope(n);
    if (res == expected)
        cout << test_name << " success." << endl;
    else
        cout << test_name << " failed." << endl;
}

int main()
{
    int n1 = 2;
    int expected1 = 1;
    test("test1", n1, expected1);

    int n2 = 10;
    int expected2 = 36;
    test("test2", n2, expected2);

    return 0;
}
